enhancements/machine-api/control-plane-machine-set.md

```diff
@@ -0,0 +1,1104 @@
+continued adoption of HA clusters.
+
+### Goals
+
+* Provide a "best practices" approach to declarative management of Control Plane Machines
+* Allow users to "1-click" scale their Control Plane to large or smaller instances
+* Allow the adoption of existing Control Planes into this new mechanism
+* Provide a safe mechanism to perform the sensitive operation of scaling the Control Plane and provide adequate
+  feedback to end users about progress of the operation
+* Allow users to opt-out of Control Plane management should our mechanism not fit their needs
+* Allow users to customise rolling update strategies based on the needs of their environments
+
+### Non-Goals
+
+* Allow horizontal scaling of the Control Plane (this may be required in the future, but today is considered
+  unsupported by OpenShift)
+* Management of the etcd cluster state (the etcd operator will handle this separately)
+* Automatic adoption of existing clusters (we will provide instructions to allow users to opt-in for existing cluster)
+* Management of anything that falls out of the scope of Machine API (e.g. management of load balancers in front of
+  Control Plane Machine - only load balancer membership is managed by Machine API)
+
+## Proposal
+
+A new CRD `ControlPlaneMachineSet` will be introduced into OpenShift and a respective `control-plane-machine-set-operator`
+will be introduced as a new second level operator within OpenShift to perform the operations described in
+this proposal.
+
+The new CRD will define the specification for managing (creating, adopting, updating) Machines for the Control Plane.
+The operator will be responsible for ensuring the desired number of Control Plane Machines are present within the
+cluster as well as providing update mechanisms akin to those seen in a StatefulSets and Deployments to allow rollouts
+of updated configuration to the Control Plane Machines.
+
+### User Stories
+
+- As a cluster administrator of OpenShift, I would like to be able to safely and automatically vertically resize my
+  control plane as and when the demand on the control plane changes
+- As a cluster administrator of OpenShift, I would like to be able to automatically recover failed Control Plane
+  Machines (eg those that have been removed by the cloud provider, or those failing MachineHealthChecks)
+- As a cluster administrator of OpenShift, I would like to be able to make changes to the configuration of the
+  underlying hardware of my control plane and have these changes safely applied using immutable hardware concepts
+- As a cluster administrator of OpenShift, I would like to be able to control rollout of changes to my Control Plane
+  Machines such that I can test changes with a single replica before applying the change to all replicas
+- (Future work) As a cluster administrator of OpenShift with restricted hardware capacity, I would like to be able to   
```

Comment:
The current approach is handling replacement w/scale up and down, the point of this line in particular is that currently we require surging the capacity as a part of the replacement. This story in particular is referencing a recreate style replacement rather than a rolling update style replacement

Reply:
That sounds like a limitation that should be explicitly expressed.

Reply:
This is already expanded upon in the `The Recreate update strategy (FUTURE WORK)` section, but I can reword this here if you think there's a way I can make this clearer in the goals

Reply:
Another note on this, this project has been designed with both cloud and local datacenter solutions in mind, though, the configuration changes we are targeting, eg resizing, don't make sense in baremetal environments. So we aren't sure whether we actually want to support this Recreate strategy ever, but we left it in for prosperity to facilitate a future conversation about this